### ISSUE TITLE
Get numbers on expired users with bills and charges

### ISSUE THREAD

**Issue by dlrueda** [sul-dlss/FOLIO-Project-Stanford#80]
Do seluser queries to get counts of expired users
CNF, CNS, CNAC
REU, RED, REG
- that have replacement charge for bills 
- that have charges (include due date)
- that have holds (active unexpired holds only, include expire date)

**Comment by jgreben** [comment]
This will inform https://github.com/sul-dlss/FOLIO-Project-Stanford/issues/77

Script and output keys and text available on `sirsi@bodoni /s/SUL/Dataload/folio_symphony_extract/Users`

**Comment by jgreben** [comment]
```
-b selects users who have less than (<), equal to (=) or greater
     than (>) the number of bills specified.
-c selects users who have less than (<), equal to (=) or greater
     than (>) the number of charges specified.
-e selects users whose privilege expires before (<), on (=), or after (>)
     the specified date.
-z selects users who have less than (<), equal to (=) or greater
     than (>) the number of available holds specified.
```

**Comment by jgreben** [comment]
CNF:
```
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user key will be read from standard input.
  The user key will be written to standard output.
  Users will be selected if the number of bills is more than 0.
  0 user record(s) considered.
  0 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user key will be read from standard input.
  The user key will be written to standard output.
  Users will be selected if the number of charges is more than 0.
  0 user record(s) considered.
  0 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user key will be read from standard input.
  The user key will be written to standard output.
  Users will be selected if the number of available holds is more than 0.
  0 user record(s) considered.
  0 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user file will be read sequentially by primary key.
  The user key will be written to standard output.
  Users will be selected if the user's privilege expires later than 2/3/2022.
  Users will be selected if user Profile is CNF.

		*** SQL ***
SELECT user_key FROM users 
WHERE (trunc(date_privilege_expires) > to_date('20220203','YYYYMMDD')) AND profile IN (22) 
ORDER BY user_key
		*** SQL ***

  357455 user record(s) considered.
  0 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
```

**Comment by jgreben** [comment]
CNS:
```
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user key will be read from standard input.
  The user key will be written to standard output.
  Users will be selected if the number of bills is more than 0.
  6 user record(s) considered.
  0 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user key will be read from standard input.
  The user key will be written to standard output.
  Users will be selected if the number of charges is more than 0.
  6 user record(s) considered.
  1 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user key will be read from standard input.
  The user key will be written to standard output.
  Users will be selected if the number of available holds is more than 0.
  6 user record(s) considered.
  0 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user file will be read sequentially by primary key.
  The user key will be written to standard output.
  Users will be selected if the user's privilege expires later than 2/3/2022.
  Users will be selected if user Profile is CNS.

		*** SQL ***
SELECT user_key FROM users 
WHERE (trunc(date_privilege_expires) > to_date('20220203','YYYYMMDD')) AND profile IN (23) 
ORDER BY user_key
		*** SQL ***

  357455 user record(s) considered.
  6 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
```

**Comment by jgreben** [comment]
CNAC:
```
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user key will be read from standard input.
  The user key will be written to standard output.
  Users will be selected if the number of bills is more than 0.
  4 user record(s) considered.
  1 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user key will be read from standard input.
  The user key will be written to standard output.
  Users will be selected if the number of charges is more than 0.
  4 user record(s) considered.
  3 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user key will be read from standard input.
  The user key will be written to standard output.
  Users will be selected if the number of available holds is more than 0.
  4 user record(s) considered.
  0 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user file will be read sequentially by primary key.
  The user key will be written to standard output.
  Users will be selected if the user's privilege expires later than 2/3/2022.
  Users will be selected if user Profile is CNAC.

		*** SQL ***
SELECT user_key FROM users 
WHERE (trunc(date_privilege_expires) > to_date('20220203','YYYYMMDD')) AND profile IN (24) 
ORDER BY user_key
		*** SQL ***

  357455 user record(s) considered.
  4 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM

```

**Comment by jgreben** [comment]
REU:
```
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user key will be read from standard input.
  The user key will be written to standard output.
  Users will be selected if the number of bills is more than 0.
  7004 user record(s) considered.
  525 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user key will be read from standard input.
  The user key will be written to standard output.
  Users will be selected if the number of charges is more than 0.
  7004 user record(s) considered.
  791 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user key will be read from standard input.
  The user key will be written to standard output.
  Users will be selected if the number of available holds is more than 0.
  7004 user record(s) considered.
  85 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user file will be read sequentially by primary key.
  The user key will be written to standard output.
  Users will be selected if the user's privilege expires later than 2/3/2022.
  Users will be selected if user Profile is REU.

		*** SQL ***
SELECT user_key FROM users 
WHERE (trunc(date_privilege_expires) > to_date('20220203','YYYYMMDD')) AND profile IN (27) 
ORDER BY user_key
		*** SQL ***

  357455 user record(s) considered.
  7004 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
```

**Comment by jgreben** [comment]
RED:
```
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user key will be read from standard input.
  The user key will be written to standard output.
  Users will be selected if the number of bills is more than 0.
  5028 user record(s) considered.
  432 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user key will be read from standard input.
  The user key will be written to standard output.
  Users will be selected if the number of charges is more than 0.
  5028 user record(s) considered.
  729 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user key will be read from standard input.
  The user key will be written to standard output.
  Users will be selected if the number of available holds is more than 0.
  5028 user record(s) considered.
  67 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user file will be read sequentially by primary key.
  The user key will be written to standard output.
  Users will be selected if the user's privilege expires later than 2/3/2022.
  Users will be selected if user Profile is RED.

		*** SQL ***
SELECT user_key FROM users 
WHERE (trunc(date_privilege_expires) > to_date('20220203','YYYYMMDD')) AND profile IN (25) 
ORDER BY user_key
		*** SQL ***

  357455 user record(s) considered.
  5028 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
```

**Comment by jgreben** [comment]
REG:
```
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user key will be read from standard input.
  The user key will be written to standard output.
  Users will be selected if the number of bills is more than 0.
  7503 user record(s) considered.
  472 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:09 PM
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:09 PM
  The user key will be read from standard input.
  The user key will be written to standard output.
  Users will be selected if the number of charges is more than 0.
  7503 user record(s) considered.
  1008 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:09 PM
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:09 PM
  The user key will be read from standard input.
  The user key will be written to standard output.
  Users will be selected if the number of available holds is more than 0.
  7503 user record(s) considered.
  110 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:09 PM
Symphony user selection 3.7.1 started on Thursday, February 3, 2022, 2:08 PM
  The user file will be read sequentially by primary key.
  The user key will be written to standard output.
  Users will be selected if the user's privilege expires later than 2/3/2022.
  Users will be selected if user Profile is REG.

		*** SQL ***
SELECT user_key FROM users 
WHERE (trunc(date_privilege_expires) > to_date('20220203','YYYYMMDD')) AND profile IN (26) 
ORDER BY user_key
		*** SQL ***

  357455 user record(s) considered.
  7503 user record(s) selected.
Symphony user selection finished on Thursday, February 3, 2022, 2:08 PM
```

**Comment by jgreben** [comment]
@dlrueda please take a look ☝️ and comment.

**Comment by dlrueda** [comment]
Couple of things

I think it needs to be ...
expire date flag -e using <  (expire date was earlier in time than 20220203)
cuz we want to see these numbers for users who have already expired.

The -b flag unfortunately is just the total number of bills this user has, not the amount on the bills.  And the total number of bills will include bills that have been paid (which we don’t care about).  I think we’ll have to use selbill and look for “amount unpaid >0” or something like that, passing it the user keys of users who have expired.
Having the bill reason and the balance would be good, so that we can have numbers for replacement bills vs overdue type bills, but I think we’d get a line for each bill.  If there’s an option to get a total unpaid amount for a user over all their unpaid bills that would be good, but not sure we can do that with a set.

Similarly, the -h flag unfortunately is only for available holds (stuff waiting on the hold shelf for the user).  We’ll need to pass user keys of expired users to selhold and look for ACTIVE holds.  Need to specify unexpired holds as well.

**Comment by jgreben** [comment]
selbill:
```
 -b selects bills with balances less than (<), equal to (=) or greater
     than (>) the specified amount.
```

selhold:
```
-j has multiple policy options [ailnov].  It selects holds with the
       specified hold status ("ACTIVE" or "INACTIVE").
```

**Comment by jgreben** [comment]
@dlrueda I updated the script and fixed the problems you identified. Please take a look at the output on `sirsi@bodoni:/s/SUL/Dataload/folio_symphony_extract/Users`

There were a bunch of `error number 111`'s for the `selhold` and `selbill` queries. Not sure what that's about, but it did otherwise find a handful of active holds and bill balances for each policy group.

**Comment by dlrueda** [comment]
I think it would be helpful to output the user key from the holds/charges/bills sels and then see how many users we are talking about, not just how many total unpaid bills are for expired users.  Same for the others.

I also would like to see the bill reason, because my memory is that we’d ignore bills that were not for replacement costs (like OVERDUE).

**Comment by dlrueda** [comment]
the error number 111s are just for users that do not have a charge (in selcharge) or a bill (in selbill).  “That user key doesn’t exist in a charge record.”  

So these are good errors, means that expired user doesn’t have an outstanding charge...

**Comment by dlrueda** [comment]
Add these profiles to scripts:
MXAC, MXF, MXAS, MXAS-6LMT, MXAS-25LMT, MXAS-BDRCT, MXD, MXFEE, MXS

And
LAWFEE, MXALUM-LAW, MXFEE-BUS

**Comment by jgreben** [comment]
Got stats for new profiles.

(script can now be run with any number of profiles given, e.g.: `./expired_user_counts.sh CNF CNS CNAC REU RED REG`)

**Comment by dlrueda** [comment]
ok, numbers look good.  I am going to close this ticket, and put final decisions for EXPIRED users in a new ticket to select from Symphony.
